cmd/scollector/collectors/aws.go

```diff
@@ -0,0 +1,314 @@
+	awsEC2StatusCheckDesc string = "The EC2 Status Check, which includes both instance-level and system-level drill-down, gathered every 60 seconds."
+	awsELBLatencyDesc     string = "The minimum, maximum and average latency as reported by the load balancer, gathered at a 60 second interval and averaged over five minutes."
+	awsELBHostCountDesc   string = "The number of instances in what the Elastic Load Balancer considers a healthy state, gathered every 60 seconds."
+)
+
+func AWS(accessKey, secretKey, region string) {
+	collectors = append(collectors, &IntervalCollector{
+		F: func() (opentsdb.MultiDataPoint, error) {
+			return c_aws(accessKey, secretKey, region)
+		},
+		Interval: 60 * time.Second,
+		name:     fmt.Sprintf("aws-%s", region),
+	})
+}
+
+func c_aws(accessKey, secretKey, region string) (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	creds := aws.Creds(accessKey, secretKey, "")
+	if creds == nil {
+		return nil, fmt.Errorf("Unable to make creds")
+	}
+	ecc := ec2.New(creds, region, nil)
+	if ecc == nil {
+		return nil, fmt.Errorf("Unable to login to EC2")
+	}
+	elb := elb.New(creds, region, nil)
+	if elb == nil {
+		return nil, fmt.Errorf("Unable to login to ELB")
+	}
+	cw := cloudwatch.New(creds, region, nil)
+	if cw == nil {
+		return nil, fmt.Errorf("Unable to login to CloudWatch")
+	}
+	instances, err := AWSGetInstances(*ecc)
+	if err != nil {
+		slog.Info("No EC2 Instances found.")
+	}
+	loadbalancers, err := AWSGetLoadBalancers(*elb)
+	if err != nil {
+		slog.Info("No ELB Load Balancecrs found.")
+	}
+	for _, loadbalancer := range loadbalancers {
+		AWSGetELBLatency(*cw, &md, loadbalancer)
+		AWSGetELBHostCounts(*cw, &md, loadbalancer)
+	}
+	for _, instance := range instances {
+		AWSGetCPU(*cw, &md, instance)
+		AWSGetNetwork(*cw, &md, instance)
+		AWSGetDiskBytes(*cw, &md, instance)
+		AWSGetDiskOps(*cw, &md, instance)
+		AWSGetStatusChecks(*cw, &md, instance)
+	}
+	return md, nil
+}
+
+func AWSGetInstances(ecc ec2.EC2) ([]ec2.Instance, error) {
+	instancelist := []ec2.Instance{}
+	resp, err := ecc.DescribeInstances(nil)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to describe EC2 Instances")
+	}
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			instancelist = append(instancelist, instance)
+		}
+	}
+	return instancelist, nil
+}
+
+func AWSGetLoadBalancers(lb elb.ELB) ([]elb.LoadBalancerDescription, error) {
+	lblist := []elb.LoadBalancerDescription{}
+	resp, err := lb.DescribeLoadBalancers(nil)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to describe ELB Balancers")
+	}
+	for _, loadbalancer := range resp.LoadBalancerDescriptions {
+		lblist = append(lblist, loadbalancer)
+	}
+	return lblist, nil
+}
+
+func AWSGetCPU(cw cloudwatch.CloudWatch, md *opentsdb.MultiDataPoint, instance ec2.Instance) error {
+	search := cloudwatch.GetMetricStatisticsInput{
+		StartTime:  time.Now().UTC().Add(time.Second * -600),
```

Comment:
CloudWatch is weird about how you get the data.  It only reports a datapoint every 5 mins unless you buy the premium per-minute monitoring.  Period can only be a multiple of 60, but it will give you a measurement every 5 minutes.  I set the time to 600 to ensure that every check produced at least one datapoint (since we could theoretically land a check at the zeroth second and get no datapoints.  However, in thinking about it, that doesn't matter too terribly because the next check would pick the datapoint up, so I should probably change this to 300 or 301.